base/filesystem.jl

```diff
@@ -192,6 +216,18 @@ function position(f::File)
     return ret
 end
 
+function eof(f::File)
+    a = Vector{UInt8}(1)
+    ret = ccall(:jl_fs_read, Int32, (Int32, Ptr{Void}, Csize_t), f.handle, a, 1)
+    systemerror("eof", ret == -1)
+    if ret == 1
+        skip(f, -1)
```

Comment:
> this seems sub-optimal
I agree, I've [replaced it](https://github.com/samoconnor/julia/commit/b2eeae0aa58a91f7708a6a771465fab882e4aec8#diff-fe07d64cf361d7f56301455da5d58544R170) with `eof(f::File) = nb_available(f) == 0`.

This is not _optimal_ either but at least it is obviously a simple default implementation (i.e. open to further optimisation).

> to really make Filesystem.File as functional and fast as IOStream, we need to  ...

As I see it there are two (or three) steps:

1) Improve test coverage of `::IO` read/write to ensure that current implementations have consistent behaviour (even if some methods for some streams are not optimised). This is important so that later optimisations for particular stream types don't break behavioural compatibility.

2) Implement a pure-Julia buffering layer on top of `File` that makes it as fast/faster than `IOStream`. I think the buffering layer should be seperate from `File` and it should work on top of any low-level IO interface.

3) (maybe) Merge `IOStream` implantation with "`BufferedFile`" to remove dependancy on `support/ios.c` At that point `support/ios.c` would be used only by flisp.